backend/rates/coingecko.go

```diff
@@ -0,0 +1,153 @@
+			if err != nil {
+				updater.log.Printf("updateHist(%v, %v): %v", start, now, err)
+				untilNext = time.Minute // TODO: exponential backoff
+			}
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+// backwardHistoryLoop updates historical market exchange rates starting with
+// the earliest fetched timestamp backwards in time.
+func (updater *RateUpdater) backwardHistoryLoop() {
+	for {
+		// When to update next, after this loop iteration is done.
+		untilNext := 10 * time.Minute // TODO: add jitter
+
+		end := updater.histEarliestTimestamp("BTC")
+		if end.IsZero() {
+			end = time.Now()
+		}
+		// 90 days is the max interval which CoinGecko responds with hourly timeseries.
+		start := end.Add(-90 * 24 * time.Hour)
+		if time.Since(start) > 720*24*time.Hour {
+			// ~2 years of historical data is enough for now.
+			// TODO: Where do we stop?
+			return
+		}
+
+		err := updater.updateHist("BTC", "USD", start, end)
+		if err != nil {
+			updater.log.Printf("updateHist(%s, %s): %v", start, end, err)
+			untilNext = time.Minute // TODO: exponential backoff
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+func (updater *RateUpdater) updateHist(coin, fiat string, start, end time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	rates, err := updater.fetchMarketRange(ctx, coin, fiat, start, end)
+	if err != nil {
+		return err
+	}
+
+	updater.histMu.Lock()
+	defer updater.histMu.Unlock()
+	// TODO: updater.hist[coin] and rates may have duplicate timestamps.
+	a := append(updater.hist[coin], rates...)
+	sort.Slice(updater.hist[coin], func(i, j int) bool {
+		return a[i].timestamp.Before(a[j].timestamp)
+	})
+	updater.hist[coin] = a
+
+	updater.Notify(observable.Event{
+		Subject: "historical_rates",
+		Action:  action.Reload,
+	})
+	return nil
+}
+
+func (updater *RateUpdater) histLatestTimestamp(coin string) time.Time {
+	updater.histMu.RLock()
+	defer updater.histMu.RUnlock()
+	var t time.Time
+	if n := len(updater.hist[coin]); n > 0 {
+		t = updater.hist[coin][n-1].timestamp
+	}
+	return t
+}
+
+func (updater *RateUpdater) histEarliestTimestamp(coin string) time.Time {
+	updater.histMu.RLock()
+	defer updater.histMu.RUnlock()
+	var t time.Time
+	if len(updater.hist[coin]) > 0 {
+		t = updater.hist[coin][0].timestamp
+	}
+	return t
+}
+
+func (updater *RateUpdater) fetchMarketRange(ctx context.Context, coin, fiat string, start, end time.Time) ([]exchangeRate, error) {
+	param := url.Values{
+		"from":        {strconv.FormatInt(start.Unix(), 10)},
+		"to":          {strconv.FormatInt(end.Unix(), 10)},
+		"vs_currency": {fiat},
+	}
+	endpoint := fmt.Sprintf("%s/coins/%s/market_chart/range?%s", updater.coingeckoURL, coin, param.Encode())
+	req, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := updater.httpClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close() //nolint:errcheck
+	var jsonBody struct {
```

Comment:
need to check `res.Status` here. as it is now, using `BTC` instead of `bitcoin`, this always returns 404, but it passes silently (even the json decoding does not fail).

Reply:
fyi there was an interesting error like this in the blockchain.info wallet: they fetched additional randomness from random.org, but some error status code, which was used as the actual randomness due to missing status code checks :sweat_smile: in combination with an additional bug, it ended up being the only "randomness" in the seed creation.

Reply:
yeah, absolutely. I forgot and made a rusty mistake here :)
nice catch!

Reply:
done